### PR TITLE
Mark lambda conversions as side-affecting

### LINKED ISSUE #22700: ArgumentException (duplicate key) in VisitLambdaOrFunction when using `+=` on an indexer with a lambda argument
**Version Used**:  2.3.2

**Steps to Reproduce**:
```csharp
using System;

public class Test {
    public void M() {
        this[() => 0] += 1;
    }
    
    public int this[Func<int> f] {
        get { return 0; }
        set {}
    }
}
```
https://sharplab.io/#v2:D4AQDABCCMDcCwAoJIDMUBMEAqBTAzgC4QDeSEFU6IALBALIAUAlKeZR4QBYCW+A2iwgBeAHwQwAXQgBqYRDjsKAXyUQ1aCDwB2xbn34gArAB4dhcQDNpZRBw4BzXMRJQA7BNgRVd+xXzOpD4cPspAA=

**Expected Behavior**:
Code is compiled correctly.

**Actual Behavior**:
```
System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Microsoft.CodeAnalysis.CSharp.LambdaRewriter.Analysis.VisitLambdaOrFunction(IBoundLambdaOrFunction node)
   at Microsoft.CodeAnalysis.CSharp.LambdaRewriter.Analysis.VisitLambda(BoundLambda node)
   at Microsoft.CodeAnalysis.CSharp.BoundLambda.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalker.VisitConversion(BoundConversion node)
   at Microsoft.CodeAnalysis.CSharp.BoundConversion.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalker.VisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalker.VisitCall(BoundCall node)
   at Microsoft.CodeAnalysis.CSharp.LambdaRewriter.Analysis.VisitCall(BoundCall node)
   at Microsoft.CodeAnalysis.CSharp.BoundCall.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalker.VisitBinaryOperator(BoundBinaryOperator node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator.VisitBinaryOperator(BoundBinaryOperator node)
   at Microsoft.CodeAnalysis.CSharp.BoundBinaryOperator.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalker.VisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalker.VisitCall(BoundCall node)
   at Microsoft.CodeAnalysis.CSharp.LambdaRewriter.Analysis.VisitCall(BoundCall node)
   at Microsoft.CodeAnalysis.CSharp.BoundCall.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalker.VisitExpressionStatement(BoundExpressionStatement node)
   at Microsoft.CodeAnalysis.CSharp.BoundExpressionStatement.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalker.VisitSequencePoint(BoundSequencePoint node)
   at Microsoft.CodeAnalysis.CSharp.BoundSequencePoint.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalker.VisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.CSharp.LambdaRewriter.Analysis.VisitBlock(BoundBlock node)
   at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.LambdaRewriter.Analysis.Analyze(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.LambdaRewriter.Rewrite(BoundStatement loweredBody, NamedTypeSymbol thisType, ParameterSymbol thisParameter, MethodSymbol method, Int32 methodOrdinal, MethodSymbol substitutedSourceMethod, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 closureDebugInfoBuilder, VariableSlotAllocator slotAllocatorOpt, TypeCompilationState compilationState, DiagnosticBag diagnostics, HashSet`1 assignLocals)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.LowerBodyOrInitializer(MethodSymbol method, Int32 methodOrdinal, BoundStatement body, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState, Boolean instrumentForDynamicAnalysis, DebugDocumentProvider debugDocumentProvider, ImmutableArray`1& dynamicAnalysisSpans, DiagnosticBag diagnostics, VariableSlotAllocator& lazyVariableSlotAllocator, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 closureDebugInfoBuilder, StateMachineTypeSymbol& stateMachineTypeOpt)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass21_0.<CompileNamedTypeAsTask>b__0()
   at Roslyn.Utilities.UICultureUtilities.<>c__DisplayClass5_0.<WithCurrentUICulture>b__0()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.WaitForWorkers()
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies(CSharpCompilation compilation, PEModuleBuilder moduleBeingBuiltOpt, Boolean emittingPdb, Boolean emitTestCoverageData, Boolean hasDeclarationErrors, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods(CommonPEModuleBuilder moduleBuilder, Boolean emittingPdb, Boolean emitMetadataOnly, Boolean emitTestCoverageData, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream metadataPEStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, CompilationTestData testData, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, Stream metadataPEStream, CancellationToken cancellationToken)
```

Note that `master` has a different (but similar) error: `System.ArgumentException: adding a duplicate`.
See: https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGA3AWACgiUBmLAJgwBU5YMBvIjFrclAFgwFkAKASkbNWImAAsAllADaAjAF4AfBjQBdDAGp5GPMJYBfPRiNkMEgHYwM4qdJQBWADwWYygGbqmhESIDmcKwYsAHYVXAxDbx8WKADGSJFI/SA===

### PR DESCRIPTION
**Pull request by agocke**
### Customer scenario

The scenario listed in #22700 demonstrates the problem. The direct cause
of the crash is that local lowering produces two instances of the lambda and 
lamda conversions in the resulting bound nodes. This crashes the closure 
conversion pass since the same lambda cannot appear in two places in the tree. 
However, even if that were allowed, this would violate the language
specification.

Placing a lambda inside an indexer which is nested in a compound
assignment creates two calls which contain the delegate produced by the
lambda expression. Since the delegate produced by the conversion can be 
compared by-reference to any other delegate produced by the same lambda, 
the code must be referentially transparent -- it must produce the same result
when evaluated repeatedly. This could be achieved by spilling the evaluation 
result into a temporary. However, the local lowering only spills if the expression
is side-affecting (which lambda conversions are not), when what it actually
needs is for the expression to be referentially transparent.

The fix is small: mark lambda conversions as side-affecting. There are
certain cases, like the creation of temporaries for out-of-order calls
with named parameters, where the lambda conversion is only required to
be non-side-affecting instead of referentially transparent, but it does
not currently seem worth the implementation effort to split up the code
path.

### Bugs this fixes

#22700

### Workarounds, if any

Manually spill the conversion into a local variable. 

### Risk

Very small, focused change. Introduces a temp, so this is more conservative than alternatives.

### Performance impact

None. No significant amount of new code.

### Is this a regression from a previous update?

No.

### Root cause analysis

Very rare scenario with very little use in the real world.

### How was the bug found?

Customer reported

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 1]
```diff
@@ -330,6 +328,8 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
                                     current = conv.Operand;
                                     break;
 
+                                case ConversionKind.AnonymousFunction:
+                                case ConversionKind.MethodGroup:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs]: >ConversionKind.MethodGroup [](start = 37, length = 26)

Let's investigate whether this change is going to have any impact on scenarios involving Linq expression trees

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs (hunk 1 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs]: I've added a similar test using an expression lambda. Please let me know if there's something else you were thinking of.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenExprLambdaTests.cs [hunk 2]
```diff
@@ -381,6 +381,42 @@ public static string GetCultureInvariantString(object value)
...
+    private static Expression<Func<int>> _f1;
+    private static Expression<Func<int>> _f2;
+
+    public static void Main() {
+        var c = new C();
+        c[() => 0] += 1;
+        Console.WriteLine(object.ReferenceEquals(_f1, _f2));
+    }
+
+    int this[Expression<Func<int>> f] {
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenExprLambdaTests.cs]: >Expression<Func<int>> [](start = 13, length = 21)

If this test is in response to my comment about expression trees, this isn't the scenario. The scenario is when the expression that uses lambda as an argument is converted to an expression tree, not just the lambda itself

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenExprLambdaTests.cs]: We now put lambda or a delegate in a temp in some cases, that might affect our ability to create an expression tree.

---
In reply to: [179239269](https://github.com/dotnet/roslyn/pull/25917#discussion_r179239269) [](ancestors = 179239269)

Comment (agocke) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenExprLambdaTests.cs]: Ah! Thanks I will try to write such a test.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenExprLambdaTests.cs]: Note that you change affects more scenarios, not just compound assignments. 

---
In reply to: [179240865](https://github.com/dotnet/roslyn/pull/25917#discussion_r179240865) [](ancestors = 179240865)

**Comment by VSadov** [comment]
The root cause of the crash is that we have code that relies on identity of bound nodes. 
I wonder if there are no other cases besides lambdas where this could happen. - I.E. if we rely on identity of a BoundLocalAccess somewhere, could we get in the same trouble?

**Comment by AlekseyTs** [comment]
> The root cause of the crash is that we have code that relies on identity of bound nodes.

Even though that is true, I believe there would be a behavior problem as we would be creating two delegates instead of one.

> I wonder if there are no other cases besides lambdas where this could happen. - I.E. if we rely on identity of a BoundLocalAccess somewhere, could we get in the same trouble?

In general, we should not be relying on identity of bound nodes because it is rather common to reuse the same nodes in multiple places in the same tree.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 3]
```diff
@@ -732,6 +743,11 @@ private static ImmutableArray<RefKind> GetRefKindsOrNull(ArrayBuilder<RefKind> r
                 }
                 refKinds[p] = argRefKind;
             }
+
+            return;
+
+            bool IsLambdaConversion(BoundExpression expr)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs]: >IsLambdaConversion [](start = 17, length = 18)

I think our convention is to start local function names with a lower-case letter.

**Comment by jaredpar** [comment]
test windows_debug_unit64_prtest  please

**Comment by jaredpar** [comment]
test windows_release_unit64_prtest please

**Comment by jaredpar** [comment]
approved

**Pull request merged**